vmm/src/interrupt.rs

```diff
@@ -241,22 +242,93 @@ impl InterruptSourceGroup for MsiInterruptGroup {
     }
 }
 
+pub struct LegacyInterruptGroup {
+    ioapic: Arc<Mutex<ioapic::Ioapic>>,
+    idx_irq_map: Mutex<HashMap<InterruptIndex, Option<u32>>>,
```

Comment:
Realistically, all legacy interrupt source groups we know about are going to be made of a single interrupt. For now, can we simply have a `irq: InterruptIndex` field instead?

We would then:

* Typically create a legacy source group with the `base_irq` being set to the actual device IRQ (e.g. 4 for the serial device) and a length of 1. The `LegacyInterruptGroup` `irq` field would then just be set to `base_irq`.
* Not support `update` for legacy interrupts.
* Have a simplified interrupt source group creation for legacy devices:
```Rust
   let device_irq = address_manager
            .allocator
            .lock()
            .unwrap()
            .allocate_irq()
            .unwrap();

   let interrupt_group = interrupt_manager
            .create_group(PIN_IRQ, device_irq as InterruptIndex, 1)
            .map_err(DeviceManagerError::CreateInterruptGroup)?;
```
* Only pass the device irq (i.e. the device source group base irq) to the device creation routine.

Reply:
That's totally fine with me since I thought about this solution. I just wanted to keep things generic, but if you think making the assumption that everything is going to be a single interrupt, then let's go for this, as this will indeed simplify :)